### PR TITLE
fix(resolving-deps-resolver): match npm's includePrerelease when hoisting peers

### LINKED ISSUE #13341: pnpm 12 auto-installs a required peer as a prerelease its declared range rejects (base-strip vs includePrerelease)
## Summary

`hoist_peers` — the required-peer picker in `pnpm/crates/resolving-deps-resolver/src/hoist_peers.rs` — accepts a prerelease version for a range that does not accept it, where the TypeScript CLI does not. It uses `satisfies_including_prerelease`, which strips the prerelease tag and retries against the base version:

```rust
pub(crate) fn satisfies_including_prerelease(range: &Range, version: &Version) -> bool {
    if range.satisfies(version) {
        return true;
    }
    if version.pre_release.is_empty() {
        return false;
    }
    let base = Version { major: version.major, minor: version.minor, patch: version.patch, .. };
    range.satisfies(&base)
}
```

The TypeScript counterpart is `semver.maxSatisfying(versions, rangeForMatch, { includePrerelease: true })` in `installing/deps-resolver/src/hoistPeers.ts`. `includePrerelease` stops semver from *excluding* prereleases of versions that fall inside the range; it does not lower the range's bound. Stripping the tag does lower it, so the two disagree whenever the candidate prerelease sits at the range's lower bound:

| version | range | strict | `includePrerelease: true` | pacquet's base-strip |
| --- | --- | --- | --- | --- |
| `18.0.0-rc.1` | `^18.0.0` | false | **false** | **true** ← diverges |
| `19.3.0-canary-28cd4bb0-20260723` | `^18.x \|\| ^19.x` | false | true | true |
| `30.0.0-alpha.6` | `^29.0.0 \|\| ^30.0.0` | false | **false** | **true** ← diverges |
| `2.0.0-alpha.1` | `^1.0.0 \|\| ^2.0.0` | false | **false** | **true** ← diverges |

The consequence is a missing required peer being auto-installed as a prerelease that the declaring package rejects. When TypeScript's `maxSatisfying` returns `null` it falls through to hoisting the *range* instead, so pnpm resolves a satisfying stable version from the registry:

```ts
} else if (isSemverRange && versions.length > 0) {
  // Preferred versions exist but none satisfies the wanted range.
  // Use the range directly so pnpm resolves it from the registry rather
  // than installing a version the peer explicitly rejects.
  if (opts.autoInstallPeers) {
    dependencies[peerName] = range
  }
}
```

pacquet has that same fallback arm — it just rarely reaches it, because the base-strip check has already claimed the prerelease as satisfying.

`pnpm/crates/resolving-deps-resolver/src/hoist_peers/tests.rs` has a test asserting the current behavior (`hoist_peers_accepts_prerelease_against_non_prerelease_range`, using exactly the `18.0.0-rc.1` / `^18.0.0` pair from the table), so the divergence is currently pinned in place by a test.

## Why this is filed rather than fixed

The Rust `node-semver` crate (2.2.0) exposes only `Range::satisfies(&Version)` — there is no `includePrerelease` option and no comparator-level access, so matching the TypeScript semantics means implementing the rule rather than passing a flag: keep prereleases eligible for comparators whose `[major, minor, patch]` tuple they share, without relaxing the range's bounds. That is a real piece of semver work with its own edge cases, and it deserves its own change and test matrix.

Note that `resolve_peers.rs` has a *third* variant, `satisfies_with_prereleases`, which is the same base-strip shape and is documented as mirroring `@yarnpkg/core`'s `satisfiesWithPrereleases` — the function the TypeScript CLI genuinely does use for peer *binding* (`resolvePeers.ts` calls `semverUtils.satisfiesWithPrereleases(…)`). That one is plausibly correct as-is; I could not verify it because `@yarnpkg/core` is not installed in my worktree. So pnpm has three different peer range checks and pacquet needs to match each to its own counterpart:

| path | TypeScript | pacquet | status |
| --- | --- | --- | --- |
| `getHoistableOptionalPeers` | `semver.satisfies` (strict) | strict | fixed (pnpm/pnpm#13320) |
| `hoistPeers` | `maxSatisfying`, `includePrerelease: true` | base-strip | **this issue** |
| `resolvePeers` | yarn `satisfiesWithPrereleases` (base-strip) | base-strip | believed correct, unverified |

## Impact

No known effect on the repositories validated so far — next.js's required-peer case is the middle row of the table, where the two agree. It surfaces when a graph carries a prerelease at a range's lower bound, e.g. an `X@2.0.0-alpha.N` pulled in by an override next to a package whose required peer is `^1 || ^2`: pacquet auto-installs the alpha where pnpm 11 resolves a stable `2.x`.

## Environment

- pnpm 12: built from `main` at `8e001491207d27af15d45da53265e38964c7737c` plus the pnpm/pnpm#13320 fixes
- pnpm 11 baseline: `11.17.0`

Found while working on pnpm/pnpm#13320 (itself from pnpm/pnpm#13305).

---
Written by an agent (Claude Code, claude-opus-5).

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

pacquet's required-peer picker (`hoist_peers`) accepted a prerelease for any range whose *base* version accepted it: it stripped the prerelease tag and retried. The TypeScript CLI uses `semver.maxSatisfying(versions, range, { includePrerelease: true })`, which is a different rule, so a `30.0.0-alpha.6` present elsewhere in the graph was auto-installed for a peer declared `^29.0.0 || ^30.0.0` — a version that peer explicitly rejects. pnpm 11 returns `null` there and falls through to hoisting the *range*, resolving a stable version from the registry.

npm's `includePrerelease` does two separate things, and only the first is what the name suggests:

1. it lifts the rule that a prerelease is eligible only for comparators carrying a prerelease of the same `major.minor.patch`;
2. it lowers to `-0` the bounds npm itself synthesized for an omitted component.

So `^18.x` becomes `>=18.0.0-0` and accepts `18.0.0-rc.1`, while the spelled-out `^18.0.0` stays `>=18.0.0` and does not. The old base-strip applied (2) unconditionally, which is where the divergence came from.

`node-semver`'s `Range::satisfies` bakes in the eligibility rule and takes no options, and its parser discards the partial/spelled-out distinction, so neither half is reachable through it. The new `IncludePrereleaseRange` evaluates each comparator on its own: endpoints still come from `node-semver`, membership goes through `Range::allows_any` against the single-version range (overlap with one point is the same endpoint test, minus the eligibility rule), and a comparator whose lower bound npm synthesized also admits the candidate's release. `<18` and `<=18` are rewritten before parsing — `node-semver` reads them as `<18.0.0` and `<=18.0.0-0` where npm reads `<18.0.0-0` and `<19.0.0-0`.

Every expectation in the new Rust test matrix is npm's own answer, read off `semver.satisfies(version, range, { includePrerelease: true })`. The implementation was validated against a generated 2652-case cross product of ranges × versions before the matrix was trimmed to the committed, readable subset.

The implementation is Rust-only — the TypeScript CLI already has the correct behavior. The TypeScript side gets the four matching test cases so the contract is pinned on both sides.

Two things deliberately left alone, both pre-existing and out of scope:

- The lockfile-reuse gate shared the base-strip helper by convenience rather than by design (it was reused from `hoist_peers` in pnpm/pnpm#12113). Its behavior is unchanged; the helper moved to its one remaining caller under the name the pattern carries elsewhere in the crate. Whether that gate should instead be strict, matching `semver.satisfies(version, wantedDep.bareSpecifier, true)` in `resolveDependencies.ts`, is a separate question.
- `node-semver` 2.2.0 reads `<=18` as `<=18.0.0-0`, which excludes `18.0.0`. That is a crate-level bug affecting every `Range::satisfies` call in pacquet, not just this one; it is worked around locally here rather than repo-wide.

Closes pnpm/pnpm#13341

## Squash Commit Body

```text
The required-peer picker accepted a prerelease for any range whose base
version accepted it, by stripping the prerelease tag and retrying. That
is not what `semver.maxSatisfying(versions, range, { includePrerelease:
true })` does in the TypeScript CLI, so a `30.0.0-alpha.6` in the graph
was auto-installed for a peer declared `^29.0.0 || ^30.0.0`, which
rejects it. pnpm 11 falls through to hoisting the range and resolves a
stable version from the registry instead.

npm's option does two separate things: it lifts the rule that a
prerelease is eligible only for comparators carrying a prerelease of the
same `major.minor.patch`, and it lowers to `-0` those bounds npm itself
synthesized for an omitted component. `^18.x` therefore starts at
`18.0.0-0` and accepts `18.0.0-rc.1`, while the spelled-out `^18.0.0`
starts at `18.0.0` and does not.

`node-semver`'s `Range::satisfies` bakes in the eligibility rule and
takes no options, and its parser drops the partial/spelled-out
distinction, so `IncludePrereleaseRange` evaluates each comparator on its
own: endpoints come from `node-semver`, membership goes through
`Range::allows_any` against the single-version range (overlap with one
point is the endpoint test without the eligibility rule), and a
comparator whose lower bound npm synthesized also admits the candidate's
release. `<18` and `<=18` are rewritten first, `node-semver` reading them
as `<18.0.0` and `<=18.0.0-0` where npm reads `<18.0.0-0` and
`<19.0.0-0`.

Expectations in the new test matrix are npm's own answers.

The lockfile-reuse gate shared the base-strip helper by convenience
rather than by design; it keeps its behavior under the name the pattern
carries elsewhere in the crate.

Closes pnpm/pnpm#13341
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users — it becomes a release note.
- [x] Added or updated tests.
- [x] Updated the documentation if needed.

---
Written by an agent (Claude Code, claude-opus-5).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/13391"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1787617475&installation_model_id=426870&pr_number=13391&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F13391&signature=c8179d7042bc8a01af2cba9534b4ac82a268f85830b3d14c4e71344ff13a7129"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved peer dependency resolution when prerelease versions are present.
  - Prevented incompatible prerelease versions from being selected for ranges that require stable releases.
  - Ensured resolution correctly handles wildcard ranges, unions, comparator ranges, and lower-bound cases.

- **Tests**
  - Added coverage for prerelease acceptance and rejection across common semver range formats.

### CODE REVIEW

**Comment by zkochan** [comment]
@coderabbitai review

---
Written by an agent (Claude Code, claude-opus-5).

File: pnpm/crates/resolving-deps-resolver/src/include_prerelease_range.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    }
+    let mut components = version.split('.');
+    let major: u64 = components.next()?.parse().ok()?;
+    let minor: Option<u64> = components.next().and_then(|minor| minor.parse().ok());
+    let patch_is_named =
+        components.next().is_some_and(|patch| patch.parse::<u64>().is_ok()) && minor.is_some();
+    match (minor, patch_is_named) {
+        (_, true) => None,
+        (None, _) => Some(format!("<{}.0.0-0", major + u64::from(inclusive))),
+        (Some(minor), _) => Some(format!("<{major}.{}.0-0", minor + u64::from(inclusive))),
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/resolving-deps-resolver/src/include_prerelease_range.rs]: Fixed in 2384dc1faa. The increment is now `checked_add`, and additionally refused once the resulting bound would exceed `node_semver::MAX_SAFE_INTEGER` — past that limit the crate rejects the rewritten comparator anyway, so producing it would be pointless.

Worth noting the end state is npm parity rather than just "no panic": npm's `Range` constructor throws on a component over `MAX_SAFE_INTEGER` and `semver.satisfies` catches that and returns `false`. Skipping the rewrite leaves the token for `node_semver`, which rejects it too, dropping the alternative — same answer. Covered by `matches_nothing_when_a_bound_names_an_out_of_range_component`, with both of the inputs you cited.

---
Written by an agent (Claude Code, claude-opus-5).

**Comment by zkochan** [comment]
Thanks — addressed the two code findings in 2384dc1faa, and pushing back on the changeset one.

**Overflow in bound rewrite (Qodo 3 / Greptile P1) — fixed.** `npm_upper_bound` now uses `checked_add` and additionally refuses a bound past `node_semver::MAX_SAFE_INTEGER`, which the crate would reject anyway. This lands on npm parity rather than merely avoiding the panic: npm throws from the `Range` constructor for an oversized component and `semver.satisfies` catches it and returns `false`; skipping the rewrite leaves the token for `node_semver`, which also rejects it and drops the alternative — same answer. New test covers both cited inputs.

**Extra parse/allocation per candidate (Qodo 2) — fixed, and the hot path is allocation-free again.** The point-range construction exists because `Range::allows_any` is the only public way to get `node_semver`'s endpoint test without its prerelease-eligibility rule. But that rule only ever *excludes prereleases* — for a release version, `Range::satisfies` already is the plain endpoint test. `satisfies` now takes that branch first, so a release candidate allocates and parses nothing, exactly as before this PR. Only prerelease candidates pay for the point ranges, which is inherent to bypassing the filter. `comparators` also switched to `Cow` so an unrewritten token is borrowed rather than copied. I re-ran the full 2652-case npm-derived matrix after the change to confirm the release fast path is exactly equivalent — all pass.

**Changeset missing a `pnpm` entry (Qodo 1) — declining.** `AGENTS.md` is explicit: *"Do not add `"pnpm"` to a Rust-only changeset: in changesets, `pnpm` always means the TypeScript CLI package. A changeset whose implementation is Rust-only and targets `pacquet` must omit `"pnpm"`."* The implementation here is Rust-only — the TypeScript CLI already had the correct behavior, and the only TS change is four added test cases pinning that contract. `@pnpm/installing.deps-resolver` publishes `files: ["lib"]`, so nothing in the published package changed and there is no release note to write for it. Qodo's own relevance note flags this rule as frequently rejected on this repo.

---
Written by an agent (Claude Code, claude-opus-5).

**Comment by zkochan** [comment]
@coderabbitai review

---
Written by an agent (Claude Code, claude-opus-5).
